src/App.tsx

```diff
@@ -0,0 +1,133 @@
+import React, { useEffect, useState } from 'react';
+import './App.css';
+import { HttpService } from './services/http.service';
+import { ISearchState, SearchComponent } from './components/search.component';
+import { ResultComponent } from './components/result.component';
+import { PaginationComponent } from './components/pagination.component';
+import { LocalStorageService } from './services/local-storage.service';
+import { IState } from './models/models';
+import { ErrorBoundaryComponent } from './components/error-boundary.component';
+import { Outlet, useLocation, useNavigate, useSearchParams } from 'react-router-dom';
+
+const initState = {
+  results: null,
+};
+
+enum UrlType {
+  init,
+  search,
+  common,
+}
+interface IPage {
+  currentPage: number;
+  itemPerPage: number;
+  searchValue: null | string;
+  type: UrlType;
+}
+export default function App() {
+  const httpService = new HttpService();
+  const localStorageService = new LocalStorageService();
+  const navigate = useNavigate();
+  const [searchParams, setSearchParams] = useSearchParams();
+  const location = useLocation();
+
+  const [mainData, setMainData] = useState<IState>({ ...initState });
+  const [url, setUrl] = useState<IPage>({
+    currentPage: 1,
+    itemPerPage: 20,
+    searchValue: localStorageService.data,
+    type: UrlType.init,
+  });
+
+  useEffect(() => {
+    const { currentPage, itemPerPage, searchValue, type } = url;
+    setMainData({ ...initState });
+    const fetchData = async () => {
+      if (type === UrlType.init && searchValue) {
+        return await httpService.search(searchValue);
+      }
+
+      if (type === UrlType.search && searchValue) {
+        localStorageService.setData(searchValue);
+        return await httpService.search(searchValue);
+      }
+      return await httpService.getData({ currentPage, itemPerPage });
+    };
+
+    fetchData().then((response) => {
+      setSearchParams(
+        (type === UrlType.search || type === UrlType.init) && searchValue
+          ? { search: searchValue }
+          : { page: String(currentPage), per_page: String(itemPerPage) }
+      );
+      setMainData({ ...mainData, results: response });
+    });
+  }, [url]);
+
+  const onQueryChange = async ({ value }: ISearchState) => {
```

Comment:
NIT: handler methods should better be start `handle`, e.g. `handleQueryChange` and props accepting this method start with `onQueryChange`